src/context/LogContext/index.tsx

```diff
@@ -31,42 +53,128 @@ const LogContextProvider: React.FC<LogContextProviderProps> = ({
   children,
   initialLogLines,
 }) => {
+  const [filters] = useQueryParam<string[]>(QueryParams.Filters, []);
+  const [bookmarks] = useQueryParam<number[]>(QueryParams.Bookmarks, []);
+  const [selectedLine] = useQueryParam<number | undefined>(
+    QueryParams.SelectedLine,
+    undefined
+  );
+  const [filterLogic] = useQueryParam(QueryParams.FilterLogic, FilterLogic.And);
+  const [upperRange] = useQueryParam<undefined | number>(
+    QueryParams.UpperRange,
+    undefined
+  );
+  const [lowerRange] = useQueryParam(QueryParams.LowerRange, 0);
   const { state, dispatch } = useLogState(initialLogLines);
+  const listRef = useRef<List>(null);
 
-  const ingestLines = useCallback(
-    (lines: string[], logType: LogTypes) => {
-      dispatch({ type: "INGEST_LOGS", logs: lines, logType });
-    },
-    [dispatch]
+  const getLine = useCallback(
+    (lineNumber: number) => state.logs[lineNumber],
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [state.logs.length]
   );
 
-  const clearLogs = useCallback(() => {
-    dispatch({ type: "CLEAR_LOGS" });
-  }, [dispatch]);
+  const scrollToLine = useCallback((lineNumber: number) => {
+    listRef.current?.scrollToRow(lineNumber);
+  }, []);
 
-  const getLine = useCallback(
-    (lineNumber: number) => state.logs[lineNumber],
-    [state.logs]
+  // TODO EVG-17537: more advanced filtering
+  const processedLogLines = useMemo(
+    () => filterLogs(state.logs, filters, bookmarks, selectedLine, filterLogic),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [state.logs.length, `${filters}`, `${bookmarks}`, selectedLine, filterLogic]
   );
 
-  const setFileName = useCallback(
-    (fileName: string) => {
-      dispatch({ type: "SET_FILE_NAME", fileName });
-    },
-    [dispatch]
+  const searchResults = useMemo(
+    () =>
+      // search through processedLoglines
+      // return the line number of the first match
+      // if no match, return undefined
+      state.searchState.searchTerm
+        ? searchLogs({
+            searchRegex: state.searchState.searchTerm,
+            processedLogLines,
+            upperBound: upperRange,
+            lowerBound: lowerRange,
+            getLine,
+          })
+        : [],
+    [
+      state.searchState.searchTerm,
+      upperRange,
+      lowerRange,
+      processedLogLines,
+      getLine,
+    ]
   );
+
+  // Handle search result changes
+  useEffect(() => {
+    dispatch({
+      type: "SET_MATCH_COUNT",
+      matchCount: searchResults.length,
+    });
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [dispatch, `${searchResults}`, state.searchState.searchTerm]);
```

Comment:
Sorry for being a `useEffect` hater... but I want to believe that we can somehow refactor this to not have to use an effect. Maybe not the best option, but it could definitely be moved to the `useMemo` for starters